### PR TITLE
Nested ldap groups

### PR DESCRIPTION
**Pull request by CarlSchwan**
This PR adds support for nested groups and also increases the amount of caching. This can be tested with https://github.com/juliushaertl/nextcloud-docker-dev/pull/51

Please look at the different commit messages for more explanation of every change

### CODE REVIEW

**Review by come-nc — COMMENTED** [review]
I am still against this solution of caching differently intermediate result instead of fixing them.
Fixing them might actually improve performances as they would be cached accross requests like other groups.

**Comment by come-nc** [comment]
> I am still against this solution of caching differently intermediate result instead of fixing them. Fixing them might actually improve performances as they would be cached accross requests like other groups.

@CarlSchwan https://github.com/nextcloud/server/commit/c21c7f2c2520fde5daf4755db1bc9b53e3c0e4d5 What do you think of this for instance?
I have 2 failing tests but they are failing before my changes as well.

**Comment by CarlSchwan** [comment]
> > I am still against this solution of caching differently intermediate result instead of fixing them. Fixing them might actually improve performances as they would be cached accross requests like other groups.
> 
> @CarlSchwan [c21c7f2](https://github.com/nextcloud/server/commit/c21c7f2c2520fde5daf4755db1bc9b53e3c0e4d5) What do you think of this for instance? I have 2 failing tests but they are failing before my changes as well.

The patch looks good and also easier to read and understand

**Comment by come-nc** [comment]
Rebased on master and solved conflicts

**Comment by CarlSchwan** [comment]
Rebased this PR and found some issues :(

With two groups that are circular: A (with 3 members) and B (with 2 members)
 
1. when loading the list of the member of A, we see 5 members, this is correct
2. when loading the list of the member of B, we only see 2 members, this is not correct
3. Clearing the redis cache
4. when loading the list of the member of B, we only see 5 members, this is correct
5. when loading the list of the member of A, we see 3 members, this is not correct

This was fixed with https://github.com/nextcloud/server/commit/682a6acd7da6ef8251c3c4430f337152a2907f77 and https://github.com/nextcloud/server/commit/4e165b1e8583c48fe8b7b8c575277d4afbad9009 in the stable20 fork

**Comment by CarlSchwan** [comment]
Found the issue! This 

```
$shouldCacheResult = count($seen) === 0;
```

got removed and when we ended up caching the intermediate steps, something that works fine for nested groups but not with circular groups. Now locally this patch seems to work.

I added it back and also added back the runtime cache optimization

File: apps/user_ldap/lib/Group_LDAP.php [hunk 1]
```diff
@@ -808,7 +808,7 @@ private function getGroupsByMember(string $dn, array &$seen = []): array {
 		if (isset($seen[$dn])) {
 			return [];
 		}
-		$cacheKey = $dn . implode('-', array_keys($seen));
+		$cacheKey = $dn . implode('-', $seen);
```
Comment (come-nc) [inline apps/user_ldap/lib/Group_LDAP.php:811]: This makes no sense, values in $seen are booleans, imploding the values cannot produce a meaningful thing.

**Comment by come-nc** [comment]
@CarlSchwan One of the points of the refactor as I remember it was to be able to cache all steps, so https://github.com/nextcloud/server/pull/30223/commits/07dec64ea991a4284743b0e20c31cd46dcfd497e looks wrong to me.

Can you provide a test case that fails without this commit?

**Comment by come-nc** [comment]
Update on this:

There was indeed a caching bug left in my code in the special case of recursive nested groups. We came to the conclusion that we have to not cache intermediate results when recursion is detected. This allows to cache all results in normal situations and only avoid caching for recursive nested groups.

This kind of bugs are not seen by tests because the tests use a mockup that does not do caching. We need to fix this and add tests that instanciate the caching so that these bugs are visible.

I have a local patch that fixes the bug but I need to coordinate with @CarlSchwan to see which part of his last commits needs reverting, and we need to test getting groups of a user, because my patch only affect getting members of a group.
Once we clear that out and have proper tests for caching, I think we can merge. We should also check what the profiler says about before VS after merging this.

**Comment by CarlSchwan** [comment]
Performance before:
- DB request: 120
- LDAP request: 169

Performance after:
- DB request: 102
- LDAP request: 103

In both case with a cold cache

**Review by come-nc — APPROVED** [review]
:+1: Tested and validated with Carl

**Review by blizzz — APPROVED** [review]

**Pull request merged**